awscli/botocore/utils.py

```diff
@@ -3305,6 +3339,231 @@ def fetch_token(
+            'grantTypes': self._AUTH_GRANT_TYPES,
+            'redirectUris': [redirect_uri],
+            'issuerUrl': issuer_url
+        }
+
+        if scopes:
+            register_kwargs['scopes'] = scopes
+        else:
+            register_kwargs['scopes'] = [self._AUTH_GRANT_DEFAULT_SCOPE]
+
+        response = self._client.register_client(**register_kwargs)
+
+        expires_at = response['clientSecretExpiresAt']
+        expires_at = datetime.datetime.fromtimestamp(expires_at, tzutc())
+        registration = {
+            'clientId': response['clientId'],
+            'clientSecret': response['clientSecret'],
+            'expiresAt': expires_at,
+            'scopes': register_kwargs['scopes'],
+            'grantTypes': register_kwargs['grantTypes']
+        }
+
+        return registration
+
+    def _registration(
+            self,
+            start_url,
+            session_name,
+            scopes,
+            force_refresh=False,
+    ):
+        cache_key = self._registration_cache_key(
+            start_url,
+            session_name,
+            scopes,
+        )
+        if not force_refresh and cache_key in self._cache:
+            registration = self._cache[cache_key]
+            if (not self._is_expired(registration) and
+                    self._is_registration_for_auth_code(registration)):
+                return registration
+
+        registration = self._register_client(
+            session_name,
+            scopes,
+            self._auth_code_fetcher.redirect_uri_without_port(),
+            start_url
+        )
+        self._cache[cache_key] = registration
+        return registration
+
+    def _extract_resolved_endpoint(self, params, **kwargs):
+        """Event handler for before-call that will extract the resolved endpoint
+        for a given request without actually running it
+        """
+        # This will contain any path and query params specific to
+        # the operation/input, so extract just the scheme and hostname
+        if params['url']:
+            parsed = urlparse(params['url'])
+            self._base_endpoint = f'{parsed.scheme}://{parsed.netloc}'
+
+        # Return a tuple containing the "response" to short-circuit the request
+        return botocore.awsrequest.AWSResponse(None, 200, {}, None), {}
+
+    def _get_base_authorization_uri(self):
+        """Simulates an SSO-OIDC request so that we can extract the "base"
+        endpoint for the current client to use for the un-modeled Authorize
+        operation
+        """
+        self._client.meta.events.register('before-call',
+                                          self._extract_resolved_endpoint)
+        self._client.register_client(
+            clientName='temp',
+            clientType='public'
+        )
+        self._client.meta.events.unregister('before-call',
+                                            self._extract_resolved_endpoint)
+
+        return self._base_endpoint
```

Comment:
For reviewers, this is so that we can resolve the "base" endpoint using the endpoint rules now that `client.meta.endpoint_url` is _not_ routed through the endpoint resolver. 

It's a little awkward since we're half-running this request, but this is possible using botocore's current public interface. 

A possibly cleaner alternative would be to expose `client._resolve_endpoint_ruleset` publicly, or allow some other public access to the SSO-OIDC endpoint resolver.